nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -848,19 +898,81 @@ ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
     ngx_http_core_run_phases(r);
 }
 
+static ngx_int_t
+ngx_http_calculate_modsec_latency(ngx_http_request_t *r, ngx_http_modsecurity_ctx_t* ctx,
+                                  ngx_http_modsecurity_loc_conf_t * cf)
+{
+    ngx_time_t *end_time;
+    ngx_msec_int_t   ms;
+    ngx_int_t ret;
+    ngx_time_update();
+    end_time = ngx_timeofday();
+
+    ms = (ngx_msec_int_t)
+        ((end_time->sec - ctx->start_time->sec) * 1000 + (end_time->msec - ctx->start_time->msec));
+    ms = ngx_max(ms, 0);
+
+    ngx_http_variable_value_t* modsec_latency_detect_var = ngx_http_get_indexed_variable(r, modsec_latency_detect_index);
+    ngx_http_variable_value_t* modsec_latency_prevent_var = ngx_http_get_indexed_variable(r, modsec_latency_prevent_index);
+    if (cf->config->is_enabled == MODSEC_DETECTION_ONLY) {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, 0);
+    }
+    else {
+        ret = ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_prevent_var, ms);
+        if (ret != NGX_OK) {
+            return ret;
+        }
+        return ngx_http_modsecurity_set_modsec_latency(r, modsec_latency_detect_var, 0);
+    }
+}
 
 /*
 ** [ENTRY POINT] does : this function called by nginx from the request handler
+*   Calculate the modsec latency in this function and invoke the main handler.
 */
 static ngx_int_t
-ngx_http_modsecurity_handler(ngx_http_request_t *r)
+ngx_http_modsecurity_handler_with_timer(ngx_http_request_t *r)
 {
+    ngx_time_update();
     ngx_http_modsecurity_loc_conf_t *cf;
-    ngx_http_modsecurity_ctx_t      *ctx;
-    ngx_int_t                        rc;
+    ngx_http_modsecurity_ctx_t *ctx;
+    ngx_int_t ret, modsec_latency_ret;
+
+    // Get module request context or create if not yet created
+    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
+    if (ctx == NULL) {
+        ctx = ngx_http_modsecurity_create_ctx(r, ngx_timeofday());
+        if (ctx == NULL) {
+            return NGX_HTTP_INTERNAL_SERVER_ERROR;
+        }
+
+        ngx_http_set_ctx(r, ctx, ngx_http_modsecurity);
+    }
 
     cf = ngx_http_get_module_loc_conf(r, ngx_http_modsecurity);
 
+    // Main modsec handler
+    ret  = ngx_http_modsecurity_handler(r, cf, ctx);
+
+    // We return failure only if memory allocation fails for latency variable in calculating metric
+    modsec_latency_ret = ngx_http_calculate_modsec_latency(r, ctx, cf);
+    if (modsec_latency_ret != NGX_OK) {
+        ngx_log_error(NGX_LOG_INFO, r->connection->log, 0, "modSecurity: latency metric memory allocation failed");
+        return modsec_latency_ret;
```

Comment:
When the alloc fails, We will return NGX_ERROR to the request processing? Is this intentional?

Reply:
> yes this is intentional. This happens when the request pool runs out of memory and we need to return NGX_ERROR to the parent. To see some references for the same go to https://msazure.visualstudio.com/One/_git/networking-roxyf-nginx?path=%2Fsrc%2Fhttp%2Fngx_http_variables.c&_a=contents&version=GBmaster, look for ngx_http_variable_request_time